cloudinit/distros/__init__.py

```diff
@@ -1248,6 +1248,13 @@ def build_dhclient_cmd(
             "/bin/true",
         ] + (["-cf", config_file, interface] if config_file else [interface])
 
+    @staticmethod
+    def eject_media(device: str) -> None:
+        cmd = ["/lib/udev/cdrom_id", "--eject-media", device]
+        if not uses_systemd():
+            cmd = ["eject", device]
+        subp.subp(cmd)
```

Comment:
@catmsred @TheRealFalcon Now we're unconditionally using this for all systemd distros. How confident are we that all systemd-based distros ship with this binary? If we don't have high confidence maybe we should fallback to "eject"?

Reply:
> It's part of upstream systemd

The presence of a tool in upstream systemd doesn't guarantee its presence on every Linux distro that runs systemd.

In Ubuntu for example, various parts of "systemd" get shipped in separate packages for various reasons. Consider:

```
Package: systemd
Package: systemd-sysv
Package: systemd-container
Package: systemd-journal-remote
Package: systemd-coredump
Package: systemd-timesyncd
Package: systemd-tests
Package: libpam-systemd
Package: libnss-myhostname
Package: libnss-mymachines
Package: libnss-resolve
Package: libnss-systemd
Package: libsystemd0
Package: libsystemd-dev
Package: libsystemd-shared
Package: udev
Package: libudev1
Package: libudev-dev
Package: udev-udeb
Package: libudev1-udeb
Package: systemd-standalone-sysusers
Package: systemd-standalone-tmpfiles
Package: systemd-oomd
Package: systemd-userdbd
Package: systemd-homed
Package: systemd-boot
Package: systemd-boot-efi
Package: systemd-resolved
```

This tool gets shipped as part of the `udev` package in Ubuntu, not the `systemd` package. I very much doubt even ubuntu minimal images are going to drop dependency on udev, but it might be reasonable to not use udev on some Ubuntu-based appliance that uses cloud-init. Honestly I'm less concerned about the Ubuntu story (I expect udev to be in use most places) and more concerned about making assumptions for all the different linux distributions that use cloud-init. Do they all ship this utility as part of some package? It is entirely possible that some distro decided not to include it in their base images because of the pre-existance of `eject`.

I would also expect `util-linux` to be available almost everywhere that the Linux kernel is (which is more ubiquitous than systemd), but in this case it was decided that it wasn't required for minimal ubuntu images. However, I could see the same argument being made but the other way. The `eject` binary is half the size of the `cdrom_id` binary on my system 43240 bytes vs 84696 bytes, respectively, so why not use `eject` in `$some_distro`'s udev rules rather than `cdrom_id`?

I'm not opposed to making the code this way, but if we are going to, I'd say we should probably take a good look at every supported distro's packaging to make sure that this is available [everywhere](https://cloudinit.readthedocs.io/en/latest/reference/distros.html). In my opinion it would be easier (and safer) to just fail over to calling `eject` when `cdrom_id` isn't called - perhaps with an INFO log level message so that future devs can collect data through bug reports about whether this ever gets used.

Reply:
Good point. We may not even want the systemd check at all. Just check for cdrom_id and if it exists, use it, otherwise use eject (or vice versa).

Reply:
Do we have a strong preference for which (eject or cd_rom) we try first?  My inclination is to try eject first since that is what has been used historically for this function.

Reply:
> My inclination is to try eject first since that is what has been used historically for this function.

Agreed.

I think I'd prefer we look before we leap when it comes to shelling out though, just because it can be a more expensive operation. Something along the lines of:
```python
if subp.which("eject"):
    call eject
elif subp.which("/lib/udev/cdrom_id"):
    call cdrom_id
else:
    raise a custom exception
```

What do you think?

Reply:
Updated raising a `subp.ProcessExecutionError`; please take a look at whether you think that's a good choice and whether the description/reason are used as they should be.